# Проект SQL

# Описание проекта
Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Первая задача — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Техническое задание

- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг — учитываются только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

# Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Импорт-библиотек-и-параметров" data-toc-modified-id="Импорт-библиотек-и-параметров-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Импорт библиотек и параметров</a></span></li><li><span><a href="#Просмотр-данных" data-toc-modified-id="Просмотр-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Просмотр данных</a></span></li><li><span><a href="#Количество-книг,-вышедших-после-1-января-2000-года" data-toc-modified-id="Количество-книг,-вышедших-после-1-января-2000-года-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Количество книг, вышедших после 1 января 2000 года</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Количество-обзоров-и-средняя-оценка-для-каждой-книги" data-toc-modified-id="Количество-обзоров-и-средняя-оценка-для-каждой-книги-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Количество обзоров и средняя оценка для каждой книги</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Автор-с-самой-высокой-средней-оценкой-книг" data-toc-modified-id="Автор-с-самой-высокой-средней-оценкой-книг-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Автор с самой высокой средней оценкой книг</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Чек-лист-доработок-" data-toc-modified-id="Чек-лист-доработок--7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>Чек-лист доработок </a></span></li></ul></li></ul></div>

## Импорт библиотек и параметров

In [1]:
# импортируем библиотеки
from IPython.display import display
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
                                                         db_config['pwd'], 
                                                         db_config['host'], 
                                                         db_config['port'],
                                                         db_config['db']) 
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# сохраняем коннектор

## Просмотр данных

In [2]:
# смотрим данные для таблицы books

query = '''
SELECT * 
FROM books
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
# смотрим данные для таблицы authors

query = '''
SELECT * 
FROM authors
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [4]:
# смотрим данные для таблицы publishers

query = '''
SELECT * 
FROM publishers
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [5]:
# смотрим данные для таблицы ratings

query = '''
SELECT * 
FROM ratings
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [6]:
# смотрим данные для таблицы reviews

query = '''
SELECT * 
FROM reviews
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Количество книг, вышедших после 1 января 2000 года

In [7]:
query = '''
SELECT 
COUNT(book_id) AS count_of_books
FROM books
WHERE publication_date > '2000-01-01'
'''
pd.io.sql.read_sql(query, con = engine)

count_of_books
0             819

### Вывод

После 1 января 2000 года вышло 819 книг

## Количество обзоров и средняя оценка для каждой книги

In [8]:
query = ''' 
SELECT 
    books.title, 
    COUNT(DISTINCT reviews.review_id) AS count_of_review, 
    AVG(rating) AS avg_rate
FROM books
    LEFT JOIN ratings on ratings.book_id = books.book_id
    LEFT JOIN reviews on reviews.book_id = books.book_id
GROUP BY 
    books.book_id
ORDER BY 
    count_of_review DESC
'''

pd.io.sql.read_sql(query, con = engine)

title  count_of_review  \
0                              Twilight (Twilight  #1)                7   
1                                  Water for Elephants                6   
2                                     The Glass Castle                6   
3    Harry Potter and the Prisoner of Azkaban (Harr...                6   
4    The Curious Incident of the Dog in the Night-Time                6   
..                                                 ...              ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel                0   
996                            The Natural Way to Draw                0   
997   The Cat in the Hat and Other Dr. Seuss Favorites                0   
998                          Essential Tales and Poems                0   
999  Disney's Beauty and the Beast (A Little Golden...                0   

     avg_rate  
0    3.662500  
1    3.977273  
2    4.206897  
3    4.414634  
4    4.081081  
..        ...  
995  3.666667  
996  3.000000  
997  5.000000  
998  4.000000  
999  4.000000  

[1000 rows x 3 columns]

### Вывод

Книга с самым большим количеством отзывов - Twilight.

## Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [9]:
query = '''
SELECT 
    publishers.publisher, 
    COUNT(book_id) AS count_of_books
FROM books 
    INNER JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE
  num_pages>50
GROUP BY
  publisher
ORDER BY
  count_of_books DESC
LIMIT(1)  
'''
top_pub = pd.io.sql.read_sql(query, con = engine)
display(top_pub)

publisher  count_of_books
0  Penguin Books              42

### Вывод

Издательство, выпустившее наибольшее число книг толще 50 страниц - Penguin Books с количеством выпущенных книг равным 42.

## Автор с самой высокой средней оценкой книг

In [10]:
query = '''
SELECT 
    author, 
    AVG(subquery.avg_rating) AS max_avg_rating
FROM 
    (SELECT 
        author,
        books.book_id,
        AVG(rating) AS avg_rating
    FROM books
        INNER JOIN authors ON books.author_id = authors.author_id
        INNER JOIN ratings ON ratings.book_id = books.book_id
    GROUP BY
        author,
        books.book_id
    HAVING
        COUNT(rating) > 50) AS subquery
GROUP BY
    author
ORDER BY
    max_avg_rating DESC
LIMIT(1)
'''
top_author = pd.io.sql.read_sql(query, con = engine)
display(top_author)

author  max_avg_rating
0  J.K. Rowling/Mary GrandPré        4.283844

### Вывод

Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré (для книг с количеством оценок больше 50).

`Я загуглила, Mary GrandPré - это иллюстратор`

## Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [11]:
query = '''
SELECT 
    AVG(review_cnt) as avg_reviews
    FROM 
        (SELECT
            COUNT(review_id) as review_cnt
            FROM
                reviews
            WHERE username IN
                (SELECT
                    username
                FROM
                    ratings
                GROUP BY
                    username
                HAVING
                    COUNT(ratings) > 50)
                GROUP BY
                    username) as sub_review_cnt 
'''
review = pd.io.sql.read_sql(query, con = engine)
display(review)

avg_reviews
0    24.333333

### Вывод

Среднее количество отзывов от пользователей, поставивших больше 50 оценок - 24,3.